libfwupdplugin/fu-cfu-offer.c

```diff
@@ -0,0 +1,561 @@
+{
+	FuCfuOfferPrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_CFU_OFFER(self));
+	priv->hw_variant = hw_variant;
+}
+
+/**
+ * fu_cfu_offer_set_protocol_revision:
+ * @self: a #FuCfuOffer
+ * @protocol_revision: integer
+ *
+ * Sets the CFU protocol version.
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_cfu_offer_set_protocol_revision(FuCfuOffer *self, guint8 protocol_revision)
+{
+	FuCfuOfferPrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_CFU_OFFER(self));
+	g_return_if_fail(protocol_revision <= 0b1111);
+	priv->protocol_revision = protocol_revision;
+}
+
+/**
+ * fu_cfu_offer_set_bank:
+ * @self: a #FuCfuOffer
+ * @bank: integer
+ *
+ * Sets bank register, used if multiple banks are supported.
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_cfu_offer_set_bank(FuCfuOffer *self, guint8 bank)
+{
+	FuCfuOfferPrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_CFU_OFFER(self));
+	g_return_if_fail(bank <= 0b11);
+	priv->bank = bank;
+}
+
+/**
+ * fu_cfu_offer_set_milestone:
+ * @self: a #FuCfuOffer
+ * @milestone: integer
+ *
+ * Sets the milestone, which can be used as a version for example EV1, EVT etc.
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_cfu_offer_set_milestone(FuCfuOffer *self, guint8 milestone)
+{
+	FuCfuOfferPrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_CFU_OFFER(self));
+	g_return_if_fail(milestone <= 0b111);
+	priv->milestone = milestone;
+}
+
+/**
+ * fu_cfu_offer_set_product_id:
+ * @self: a #FuCfuOffer
+ * @product_id: integer
+ *
+ * Sets the product ID for this CFU image.
+ *
+ * Since: 1.7.0
+ **/
+void
+fu_cfu_offer_set_product_id(FuCfuOffer *self, guint16 product_id)
+{
+	FuCfuOfferPrivate *priv = GET_PRIVATE(self);
+	g_return_if_fail(FU_IS_CFU_OFFER(self));
+	priv->product_id = product_id;
+}
+
+static gboolean
+fu_cfu_offer_parse(FuFirmware *firmware,
+		   GBytes *fw,
+		   guint64 addr_start,
+		   guint64 addr_end,
+		   FwupdInstallFlags flags,
+		   GError **error)
+{
+	FuCfuOffer *self = FU_CFU_OFFER(firmware);
+	FuCfuOfferPrivate *priv = GET_PRIVATE(self);
+	gsize bufsz = 0;
+	guint8 tmp = 0;
+	guint32 tmp32 = 0;
+	const guint8 *buf = g_bytes_get_data(fw, &bufsz);
+
+	/* component info */
+	if (!fu_common_read_uint8_safe(buf, bufsz, 0x0, &priv->segment_number, error))
+		return FALSE;
+	if (!fu_common_read_uint8_safe(buf, bufsz, 0x1, &tmp, error))
+		return FALSE;
+	priv->force_ignore_version = (tmp & 0b1) > 0;
+	priv->force_immediate_reset = (tmp & 0b10) > 0;
+	if (!fu_common_read_uint8_safe(buf, bufsz, 0x2, &priv->component_id, error))
+		return FALSE;
+	if (!fu_common_read_uint8_safe(buf, bufsz, 0x3, &priv->token, error))
+		return FALSE;
+
+	/* version */
+	if (!fu_common_read_uint32_safe(buf, bufsz, 0x4, &tmp32, G_LITTLE_ENDIAN, error))
+		return FALSE;
+	fu_firmware_set_version_raw(firmware, tmp32);
+	if (!fu_common_read_uint32_safe(buf, bufsz, 0x8, &priv->hw_variant, G_LITTLE_ENDIAN, error))
+		return FALSE;
+
+	/* product info */
```

Comment:
aren't there vendor specific bytes that go inbetween?

Reply:
It's technically *reserved*.

	typedef struct
	{
		struct
		{
			UINT8 segmentNumber;
			UINT8 reserved0 : 6;
			UINT8 forceImmediateReset : 1;
			UINT8 forceIgnoreVersion : 1;
			UINT8 componentId;
			UINT8 token;
		} componentInfo;
		UINT32 version;
		UINT32 hwVariantMask;
		struct
		{
			UINT8 protocolRevision : 4;
			UINT8 bank : 2;
			UINT8 reserved0 : 2;
			UINT8 milestone : 3;
			UINT8 reserved1 : 5;
			UINT16 productId;
		} productInfo;
	} FWUPDATE_OFFER_COMMAND;

I think if if the vendor is using the reserved area they shouldn't be, but if they are the new-plugin can superclass FuCfuOffer into something else and then call the parent->parse vfunc after/before the custom parsing.

Reply:
OK sounds good.